src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
@juherr - The test that I added as part of this commit will essentially fail and can be considered as a test for `DynamicGraph`. On a side note, this is just a hunch. I don't think `DynamicGraph` basically distinguishes edges with the same weight on their priorities. I have fixed this with an additional commit.

> Yes and no. We can suppose they will be finished and won't take time. But we can imagine a more elegant solution.
I would like to understand the reason behind why you feel that the `Latch` based approach is not a befitting solution. Can you please help shed some light around that ?